src/core/lib/gpr/subprocess.cc

```diff
@@ -0,0 +1,471 @@
+        // We're done reading.  Close.
+        CloseHandleOrDie(child_stdout_);
+        child_stdout_ = nullptr;
+      } else {
+        output_data->append(buffer, n);
+      }
+    }
+  }
+
+  if (child_stdin_ != nullptr) {
+    // Child did not finish reading input before it closed the output.
+    // Presumably it exited with an error.
+    CloseHandleOrDie(child_stdin_);
+    child_stdin_ = nullptr;
+  }
+
+  DWORD wait_result = WaitForSingleObject(child_handle_, INFINITE);
+
+  if (wait_result == WAIT_FAILED) {
+    ABSL_LOG(FATAL) << "WaitForSingleObject: "
+                    << Win32ErrorMessage(GetLastError());
+  } else if (wait_result != WAIT_OBJECT_0) {
+    ABSL_LOG(FATAL) << "WaitForSingleObject: Unexpected return code: "
+                    << wait_result;
+  }
+
+  DWORD exit_code;
+  if (!GetExitCodeProcess(child_handle_, &exit_code)) {
+    ABSL_LOG(FATAL) << "GetExitCodeProcess: "
+                    << Win32ErrorMessage(GetLastError());
+  }
+
+  CloseHandleOrDie(child_handle_);
+  child_handle_ = nullptr;
+
+  if (exit_code != 0) {
+    *error = absl::Substitute("Plugin failed with status code $0.", exit_code);
+    return false;
+  }
+
+  return true;
+}
+
+std::string Subprocess::Win32ErrorMessage(DWORD error_code) {
+  char* message;
+
+  // WTF?
+  FormatMessageA(FORMAT_MESSAGE_ALLOCATE_BUFFER | FORMAT_MESSAGE_FROM_SYSTEM |
+                     FORMAT_MESSAGE_IGNORE_INSERTS,
+                 nullptr, error_code,
+                 MAKELANGID(LANG_ENGLISH, SUBLANG_ENGLISH_US),
+                 (LPSTR)&message,  // NOT A BUG!
+                 0, nullptr);
+
+  std::string result = message;
+  LocalFree(message);
+  return result;
+}
+
+// ===================================================================
+
+#else  // _WIN32
+
+Subprocess::Subprocess()
+    : child_pid_(-1), child_stdin_(-1), child_stdout_(-1) {}
+
+Subprocess::~Subprocess() {
+  if (child_stdin_ != -1) {
+    close(child_stdin_);
+  }
+  if (child_stdout_ != -1) {
+    close(child_stdout_);
+  }
+}
+
+void Subprocess::Start(const std::string& command,
+                       std::vector<std::string> envp_vector) {
+  // Note that we assume that there are no other threads, thus we don't have to
+  // do crazy stuff like using socket pairs or avoiding libc locks.
+
+  // [0] is read end, [1] is write end.
+  int stdin_pipe[2];
+  int stdout_pipe[2];
+
+  int p0 = pipe(stdin_pipe);
+  int p1 = pipe(stdout_pipe);
+  GPR_ASSERT(p0 != -1);
+  GPR_ASSERT(p1 != -1);
+
+  std::vector<std::string> arg_vector = absl::StrSplit(command, " ");
+  char** argv = (char**)malloc(sizeof(char*) * (arg_vector.size() + 1));
+  for (int i = 0; i < arg_vector.size(); i++) {
+    argv[i] = portable_strdup(arg_vector[i].c_str());
+  }
+  argv[arg_vector.size()] = nullptr;
+  int32_t environ_count = 0, i = 0;
+  while (environ[environ_count] != nullptr) environ_count += 1;
+  char** envp =
+      (char**)malloc(sizeof(char*) * (environ_count + envp_vector.size() + 1));
```

Comment:
Can you put this logic in the pluggable auth class, and then have the subprocess just call with whatever environment variables are provided? We want to use the existing environ variables but not every use case would.